sls_api/endpoints/tools/files.py

```diff
@@ -443,6 +448,216 @@ def get_file_tree(project, file_path=None):
+
+    # Process the XML file
+    metadata, error_message, status_code = extract_publication_metadata_from_tei_xml(full_path)
+    if error_message:
+        return create_error_response(error_message, status_code=status_code)
+    else:
+        return create_success_response("Metadata retrieved from XML file.", data=metadata)
+
+
+def extract_publication_metadata_from_tei_xml(file_path: str) -> Tuple[Optional[Dict[str, Any]], Optional[str], Optional[int]]:
+    """
+    Extracts publication metadata (document title, date of origin and main
+    language) from a TEI XML file located at the given file path.
+
+    Args:
+
+        file_path (str): The absolute path to the TEI XML file.
+
+    Returns:
+
+    - A tuple containing:
+        - metadata (dict or None): A dictionary with the extracted
+          metadata:
+            - "name" (str or None): The title extracted from the XML.
+            - "original_publication_date" (str or None): The date of
+              origin in "YYYY", "YYYY-MM", or "YYYY-MM-DD" format.
+            - "language" (str or None): The language code.
+            - "genre" (str or None): The genre of the text.
+            Returns `None` if an error occurred.
+        - error_message (str or None): An error message if an error
+          occurred; otherwise `None`.
+        - status_code (int or None): The HTTP status code corresponding to
+        the result (e.g., 200 on success, 404 if file not found).
+
+    Examples:
+
+        >>> metadata, error_message, status_code = extract_publication_metadata_from_tei_xml('/path/to/file.xml')
+    """
+    try:
+        # Parse the XML file and extract relevant metadata from it
+        with open(file_path, "r", encoding="utf-8-sig") as xml_file:
+            tree = ET.parse(xml_file)
+        root = tree.getroot()
+
+        # Determine namespace
+        ns = {'tei': 'http://www.tei-c.org/ns/1.0'}
+
+        # Helper function to get full text including subelements
+        def get_full_text(element):
+            return "".join(element.itertext()) if element is not None else None
+
+        # Extract the full text of <title> inside <titleStmt>
+        title_element = root.find("./tei:teiHeader/tei:fileDesc/tei:titleStmt/tei:title", namespaces=ns)
+        title = get_full_text(title_element)
+
+        # Extract the @when attribute value in <origDate> within <sourceDesc>
+        orig_date_element = root.find("./tei:teiHeader/tei:fileDesc/tei:sourceDesc//tei:origDate", namespaces=ns)
+        orig_date = orig_date_element.get("when") if orig_date_element is not None else None
+        if not orig_date:
+            # Search for a <date> with @when in <bibl> within <sourceDesc>
+            date_element = root.find("./tei:teiHeader/tei:fileDesc/tei:sourceDesc/tei:bibl//tei:date", namespaces=ns)
+            orig_date = date_element.get("when") if date_element is not None else None
+
+            # Validate orig_date, must conform to YYYY, YYYY-MM
```

Comment:
A cleaner way to handle this would be to have a wrapper method `is_valid_date` that checks if the date is one of the acceptable formats, so that this function doesn't need to check thrice, but this is likely also fine.

Reply:
Added a wrapper function for the check.